libsql/src/local/database.rs

```diff
@@ -380,84 +387,54 @@ impl Database {
     #[cfg(feature = "sync")]
     /// Push WAL frames to remote.
     pub async fn push(&self) -> Result<crate::database::Replicated> {
-        let sync_ctx = self.sync_ctx.as_ref().unwrap();
+        let mut ctx = match &self.sync_ctx {
+            Some(ctx) => ctx.lock().await,
+            None => panic!("sync context not set"),
+        };
+
         let conn = self.connect()?;
 
+        // TODO: can this be cached?
         let page_size = {
-            let rows = conn.query("PRAGMA page_size", crate::params::Params::None)?.unwrap();
+            let rows = conn
+                .query("PRAGMA page_size", crate::params::Params::None)?
+                .unwrap();
             let row = rows.next()?.unwrap();
             let page_size = row.get::<u32>(0)?;
             page_size
         };
 
-        let mut max_frame_no: std::os::raw::c_uint = 0;
-        unsafe { libsql_sys::ffi::libsql_wal_frame_count(conn.handle(), &mut max_frame_no) };
-        
+        let max_frame_no = conn.wal_frame_count();
+
         let generation = 1; // TODO: Probe from WAL.
-        let start_frame_no = sync_ctx.durable_frame_num + 1;
+        let start_frame_no = ctx.durable_frame_num() + 1;
         let end_frame_no = max_frame_no;
 
+        // TODO: figure out relation to durable_frame_num
+        // let max_frame_no = ctx.max_frame_no();
```

Comment:
@penberg before we can merge this would be helpful to clairify the difference between durable_frame_ num and max_frame_no. What I understand is we want to write the durable_frame_num here (which I guess is the max_frame_no from the server) to disk to ensure we start at a more advanced wal frame than 0.

Reply:
You always have two different max frame numbers: one on client, one on server. I tend to call the latter `durable_frame_num`.